File: 328-odd-even-linked-list/328-odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode *evenS = NULL, *evenE=NULL, *oddS=NULL, *oddE=NULL;
            int data=0;
            for(ListNode *curr=head;curr!=NULL;curr=curr->next){
                    // int data = curr->val;
                    if(data%2==0){
                            if(evenS==NULL){
                                    evenS=curr;
                                    evenE=evenS;
                                    data++;
                            }
                            else{
                                    evenE->next = curr;
                                    evenE=evenE->next;
                                    data++;
                            }
                    }
                    else{
                           if(oddS==NULL){
                                    oddS=curr;
                                    oddE=oddS;
                                   data++;
                            }
                            else{
                                    oddE->next = curr;
                                    oddE=oddE->next;
                                    data++;
                            } 
                    }
            }
            if(oddS==NULL || evenS==NULL)
                    return head;
            evenE->next = oddS;
            oddE->next = NULL;
            
            return evenS;
    }
};

File: 328-odd-even-linked-list/328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        ListNode evenS = null, evenE=null, oddS=null, oddE=null;
            int index=1;
            for(ListNode curr =  head; curr != null; curr = curr.next){
                    if (index % 2 == 0) {
                            if(evenS==null){
                                    evenS=curr;
                                    evenE=evenS;
                                    index++;
                            }
                            else{
                                    evenE.next = curr;
                                    evenE=evenE.next;
                                    index++;
                            }
                    }
                    else{
                           if(oddS==null){
                                    oddS=curr;
                                    oddE=oddS;
                                      index++;
                            }
                            else{
                                    oddE.next = curr;
                                    oddE=oddE.next;
                                    index++;
                            } 
                    }
            }
            if(oddS==null || evenS==null)
                    return head;
            oddE.next = evenS;
            evenE.next = null;
            
            return oddS;
    }
}